In [ ]:
!pip install surprise

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 154.4/154.4 kB 1.8 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  Created wheel for scikit-surprise: filename=scikit_surprise-1.1.4-cp312-cp312-linux_x86_64.whl size=2610392 sha256=2e8851f0bc81d66f15af81b5cd12ead04474ce0900c71d292518ec91257016fb
  Stored in directory: /root/.cache/pip/wheels/75/fa/bc/739bc2cb1fbaab6061854e6cfbb81a0ae52c92a502a7fa454b
Successfully built scikit-surprise


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from sklearn.metrics.pairwise import linear_kernel, cosine_similarity
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, mean_absolute_error
from surprise import Reader, Dataset, SVD
from surprise.model_selection import cross_validate, train_test_split as surprise_split
import warnings
warnings.filterwarnings('ignore')

###Load and Prepare Data

In [ ]:
credits_df = pd.read_csv('/content/drive/MyDrive/Case Studies/Netflix/datasets/tmdb_5000_credits.csv')
movies_df = pd.read_csv('/content/drive/MyDrive/Case Studies/Netflix/datasets/tmdb_5000_movies.csv')

In [ ]:
credits_df.columns = ['id','title','cast','crew']
movies_df = movies_df.merge(credits_df, on='id')

In [ ]:
ratings_df = pd.read_csv('/content/drive/MyDrive/Case Studies/Netflix/datasets/ratings_small.csv')

###Engineer Features

In [ ]:
#Comprehensive feature engineering for business insights
from ast import literal_eval

movies_df['overview'] = movies_df['overview'].fillna('')

In [ ]:
#Parse the features into corresponding Python objects
json_features = ['cast', 'crew', 'keywords', 'genres']

for feature in json_features:
  if feature in movies_df.columns:
    movies_df[feature] = movies_df[feature].apply(lambda x: literal_eval(x) if pd.notna(x) else [])

In [ ]:
#Extract director's name from the crew feature
def get_director(crew_list):
  if isinstance(crew_list,list):
    for person in crew_list:
      if person['job'] == 'Director':
        return person['name']
    return np.nan

movies_df['director'] = movies_df['crew'].apply(get_director)

In [ ]:
#Extract top 3 cast and features from a list
def get_top_items(item_list):
  if isinstance(item_list,list):
    names = [item['name'] for item in item_list]

    if len(names) > 3:
      names = names[:3]
    return names

  return []


features = ['cast','keywords','genres']
for feature in features:
  if feature in movies_df.columns:
    movies_df[feature] = movies_df[feature].apply(get_top_items)


In [ ]:
#Building some other business relevant features
movies_df['release_date'] = pd.to_datetime(movies_df['release_date'])
movies_df['release_year'] = pd.to_datetime(movies_df['release_date'],errors='coerce').dt.year

movies_df['roi'] = (movies_df['revenue'] / (movies_df['budget'] + 1)).fillna(0)
movies_df['is_recent'] = (movies_df['release_year'] >= 2015).astype(int)

###Clean Data

In [ ]:
#Clean and standardize text features
def clean_text(text_list):
  if isinstance(text_list, list):
    # Ensure each item is a string before applying lower and replace
    return [str(item).lower().replace(" ", "") for item in text_list if pd.notna(item)]
  elif isinstance(text_list, str):
    return str(text_list).lower().replace(" ", "")
  # Return an empty list for NaN or other unexpected types in list columns
  return []

features = ['cast','keywords','director','genres']
for feature in features:
  if feature in movies_df.columns:
    movies_df[feature] = movies_df[feature].apply(clean_text)

In [ ]:
#Create a metadata soup
# i.e. a string that contains all the metadata that we want to feed our vectorizer(namely actors,directors and keywords)

def create_content_soup(row):
  features=[]

  if isinstance(row['keywords'],list):
    features.extend(row['keywords']*2)
  if isinstance(row['cast'],list):
    features.extend(row['cast'])
  if isinstance(row['genres'],list):
    features.extend(row['genres']*2)
  if isinstance(row['director'],str):
    features.extend([row['director']]*3)

  return ' '.join(features)

movies_df['content_features'] = movies_df.apply(create_content_soup,axis=1)

###Calculate Weighted Rating

In [ ]:
#define some business rules beforehand
business_rules = {
    'min_votes_threshold':50,
    'popularity_weight': 0.3,
    'recency_weight': 0.2,
    'diversity_factor': 0.15
}

In [ ]:
# Calculate Imdb style weighted ratings with business considerations
# Higher threshold = more exclusive recommendations

C = movies_df['vote_average'].mean()
m = movies_df['vote_count'].quantile(0.8)

def weighted_score(row, C=C, m=m):
  v = row['vote_count']
  R = row['vote_average']

  base_score = (v /(v+m)*R) + (m/(v+m)*C)

  popularity_boost = min(row['popularity']/100,1) *  business_rules['popularity_weight']
  recency_boost = row['is_recent'] * business_rules['recency_weight']

  return base_score + popularity_boost + recency_boost

movies_df['weighted_score'] = movies_df.apply(weighted_score,axis=1)

In [ ]:
#Movies that qualify for recommendations

qualified_movies = len(movies_df[movies_df['vote_count'] >= m])

###Building Content-based recommendation engine

In [ ]:
#Use CountVectorizer for cast/crew (don't penalize frequent collaborations)
count_vectorizer = CountVectorizer(stop_words='english',max_features=10000)
content_matrix = count_vectorizer.fit_transform(movies_df['content_features'])

In [ ]:
#Calculate similarity matrix
content_sim_matrix = cosine_similarity(content_matrix,content_matrix)

#Create reverse index mapping
indices = pd.Series(movies_df.index, index = movies_df['title_x']).drop_duplicates()

###Building Collaborative Engine

In [ ]:
#Prepare data for Surprise library
reader = Reader(rating_scale=(0.5,5.0))
data = Dataset.load_from_df(ratings_df[['userId','movieId','rating']],reader)

In [ ]:
#traintest split with temporal conditions
trainset, testset = surprise_split(data,test_size=0.2,random_state=42)

In [ ]:
#Train SVD model
svd_model = SVD(
    n_factors=50, #latent factors
    n_epochs=20, #number of iterations
    lr_all=0.005, #learning rate
    reg_all=0.02, #regularization
    random_state=42
)

svd_model.fit(trainset)

In [ ]:
#Evaluate the model
predictions = svd_model.test(testset)
rmse = np.sqrt(mean_squared_error(
    [pred.r_ui for pred in predictions],
    [pred.est for pred in predictions]
))

print(f"RMSE: {rmse:.3f}")

RMSE: 0.899


###Demographic Recommendations

In [ ]:
#Get top-rated movies with demographic filtering
n_recommendations = 10

top_movies = movies_df.nlargest(n_recommendations,'weighted_score')

top_movies[['title_x','vote_average','vote_count','weighted_score','genres']].copy()

,title_x,vote_average,vote_count,weighted_score,genres
1881,The Shawshank Redemption,8.5,8205,8.548353,"[drama, crime]"
662,Fight Club,8.3,9413,8.396134,[drama]
3337,The Godfather,8.4,5893,8.377404,"[drama, crime]"
3232,Pulp Fiction,8.3,8428,8.374738,"[thriller, crime]"
65,The Dark Knight,8.2,12002,8.344250,"[drama, action, crime]"
809,Forrest Gump,8.2,7927,8.272814,"[comedy, drama, romance]"
96,Inception,8.1,13752,8.269290,"[action, thriller, sciencefiction]"
77,Inside Out,8.0,6560,8.256979,"[drama, comedy, animation]"
95,Interstellar,8.1,10867,8.237399,"[adventure, drama, sciencefiction]"
1818,Schindler's List,8.3,4329,8.200080,"[drama, history, war]"


###Content-based recommendations

In [ ]:
#Get content-based recommendations with business logic
def get_content_recommendations(movie_title,n_recommendations=10):
  try:
    movie_idx = indices[movie_title]

    sim_scores = list(enumerate(content_sim_matrix[movie_idx]))
    sim_scores = sorted(sim_scores,key = lambda x:x[1],reverse=True)

    similar_indices = [i[0] for i in sim_scores[1:n_recommendations+1]]

    recommendations = movies_df.iloc[similar_indices].copy()
    recommendations['similarity_score'] = [i[1] for i in sim_scores[1:n_recommendations+1]]

    return recommendations[['title_x','genres','director','vote_average','similarity_score']]

  except KeyError:
    available_titles = movies_df['title_x'].str.contains(movie_title,case=False, na=False)
    suggestions = movies_df[available_titles]['title_x'].head(5).tolist()

    return f"Movie {movie_title} not found, did you mean: {suggestions}"

###Collaborative Recommendations

In [ ]:
# Function to get collaborative recommendations using SVD from Surprise
def get_collaborative_recommendations(user_id, n_recommendations=10):
    try:
        # Get a list of all movie ids
        all_movie_ids = movies_df['id'].unique()

        # Get movies the user has already rated
        user_rated_movie_ids = ratings_df[ratings_df['userId'] == user_id]['movieId'].tolist()

        # Predict ratings for movies the user hasn't rated
        unrated_movie_ids = [movie_id for movie_id in all_movie_ids if movie_id not in user_rated_movie_ids]

        predictions = [svd_model.predict(user_id, movie_id) for movie_id in unrated_movie_ids]

        # Sort predictions by estimated rating
        predictions.sort(key=lambda x: x.est, reverse=True)

        # Get the top recommended movie ids
        top_recommended_movie_ids = [pred.iid for pred in predictions[:n_recommendations]]

        # Get details of the recommended movies
        recommendations = movies_df[movies_df['id'].isin(top_recommended_movie_ids)].copy()

        # Add predicted ratings to the recommendations (optional)
        # This requires mapping the predicted ratings back to the dataframe
        predicted_ratings = {pred.iid: pred.est for pred in predictions[:n_recommendations]}
        recommendations['predicted_rating'] = recommendations['id'].map(predicted_ratings)


        return recommendations[['title_x', 'genres', 'vote_average', 'predicted_rating']]

    except ValueError:
        return f"User ID {user_id} not found in the training data."

###Hybrid Recommendations

In [ ]:
#Advanced Hybrid Recommendations combining multiple approaches
# main production method that handles cold start and business logic

def get_hybrid_recommendations(user_id=None,movie_title=None,n_recommendations=10):
  recommendations = pd.DataFrame()

  #if user exists and has ratings
  if user_id and ratings_df is not None:
    user_ratings_count = len(ratings_df[ratings_df['userId']==user_id])

    if user_ratings_count >= 5:
      collab_recs = get_collaborative_recommendations(user_id,n_recommendations) # This line calls the updated function
      if isinstance(collab_recs,pd.DataFrame):
        collab_recs['recommendation_type'] = 'Collaborative'
        recommendations = collab_recs

  #content-based for specific movie interest
  if movie_title and len(recommendations) < n_recommendations:
    content_recs = get_content_recommendations(movie_title,n_recommendations//2)
    if isinstance(content_recs,pd.DataFrame):
      content_recs['recommendation_type'] = 'Content-Based'
      recommendations =  pd.concat([recommendations,content_recs], ignore_index=True)

  #Demographic fallback (for cold start)
  if len(recommendations) < n_recommendations:
    demo_recs = get_demographic_recommendations(n_recommendations)
    demo_recs['recommendation_type'] = 'Trending'
    recommendations = pd.concat([recommendations,demo_recs], ignore_index=True)

  recommendations = recommendations.drop_duplicates(subset=['title_x']).head(n_recommendations)

  return recommendations

###Evaluate System Performance

In [ ]:
#Comprehensive evaluation of the recommendation system

n_users = ratings_df['userId'].nunique()
n_movies = ratings_df['movieId'].nunique()
n_ratings = len(ratings_df)
sparsity = 1 - (n_ratings / (n_users * n_movies))

print(f"   Users: {n_users:,}")
print(f"   Movies: {n_movies:,}")
print(f"   Ratings: {n_ratings:,}")
print(f"   Sparsity: {sparsity:.3%}")

   Users: 671
   Movies: 9,066
   Ratings: 100,004
   Sparsity: 98.356%


In [ ]:
#Collaboative Filtering Performance

if svd_model:
  reader = Reader(rating_scale=(0.5,5.0))
  data = Dataset.load_from_df(ratings_df[['userId','movieId','rating']], reader)

  cv_results = cross_validate(svd_model, data, measures=['RMSE','MAE'],cv=3,verbose=False)
  print(f"RMSE: {cv_results['test_rmse'].mean():.3f} (±{cv_results['test_rmse'].std():.3f})")
  print(f"MAE: {cv_results['test_mae'].mean():.3f} (±{cv_results['test_mae'].std():.3f})")

RMSE: 0.900 (±0.010)
MAE: 0.693 (±0.006)


###Generate Business Insights

In [ ]:
#Business Metrics
popular_genres = movies_df['genres'].apply(lambda x: x[0] if isinstance(x, list) and len(x) > 0 else 'Unknown').value_counts().head(3)
print(f"Top genres: {', '.join(popular_genres.index)}")

high_roi_movies = movies_df[movies_df['roi'] > 3]['title_x'].count()
print(f"Movies with High ROI (>3x): {high_roi_movies}")

Top genres: drama, comedy, action
Movies with High ROI (>3x): 1421


In [ ]:
#Content strategy insights

avg_rating_by_genre = {}
revenue_by_genre = {}

for _,movie in movies_df.iterrows():
  # Check if genres list is not empty before iterating
  if isinstance(movie['genres'], list) and len(movie['genres']) > 0:
    primary_genre = movie['genres'][0]
    if primary_genre not in avg_rating_by_genre:
      avg_rating_by_genre[primary_genre] = []
      revenue_by_genre[primary_genre] = []

    avg_rating_by_genre[primary_genre].append(movie['vote_average'])
    if pd.notna(movie['revenue']):
      revenue_by_genre[primary_genre].append(movie['revenue'])

In [ ]:
#Calculate averages
genre_insights = {}
for genre in avg_rating_by_genre:
  genre_insights[genre] = {
      'avg_rating': np.mean(avg_rating_by_genre[genre]),
      'avg_revenue': np.mean(revenue_by_genre[genre]),
      'movie_count': len(avg_rating_by_genre[genre])
  }

In [ ]:
#Top performing genres
top_rated_genres = sorted(genre_insights.items(), key=lambda x: x[1]['avg_rating'], reverse=True)[:3]
top_revenue_genres = sorted(genre_insights.items(), key=lambda x: x[1]['avg_revenue'], reverse=True)[:3]

print(f"Highest Rated Genres: {', '.join([g[0] for g in top_rated_genres])}")
print(f"Highest Revenue Genres: {', '.join([g[0] for g in top_revenue_genres])}")

Highest Rated Genres: foreign, history, war
Highest Revenue Genres: animation, adventure, sciencefiction


In [ ]:
#User engagement patterns
if ratings_df is not None:
  avg_user_ratings = ratings_df.groupby('userId')['rating'].count().mean()
  active_users = ratings_df.groupby('userId')['rating'].count() # Get rating count for each user
  active_users = active_users[active_users >= avg_user_ratings].nunique() # Count unique users who meet the criteria

  print(f"Average ratings per user: {avg_user_ratings:.1f}")
  print(f"Highly Engaged Users: {active_users/ratings_df['userId'].nunique():.1%}") # Modified to only show percentage

Average ratings per user: 149.0
Highly Engaged Users: 22.2%


###Recommendation Report

In [ ]:
#Generate a comprehensive recommendation report

def create_recommendation_report(user_id=None, movie_title=None, save_report=True):

  recommendations = get_hybrid_recommendations(user_id, movie_title)
  if user_id:
    print(f"   Target User: {user_id}")
  if movie_title:
    print(f"   Based on Interest: {movie_title}")

  for idx, movie in recommendations.iterrows():
    print(f"{idx+1}. {movie['title_x']}")
    if 'genres' in movie and movie['genres']:
      print(f"   Genres: {', '.join(movie['genres'][:3])}")
    if 'vote_average' in movie:
      print(f"   Rating: {movie['vote_average']:.1f}/10")
    if 'recommendation_type' in movie:
      print(f"   Strategy: {movie['recommendation_type']}")
      print()

  return recommendations

In [ ]:
# Demographic fallback function
def get_demographic_recommendations(n_recommendations=10):
    """Gets top N movies based on weighted score for demographic recommendations."""
    top_movies = movies_df.nlargest(n_recommendations,'weighted_score')
    return top_movies[['title_x', 'genres', 'vote_average', 'weighted_score']].copy()

In [ ]:
# Example usage of the recommendation report function
# Replace with a user_id and/or movie_title to get specific recommendations
# create_recommendation_report(user_id=1, movie_title='Avatar')
# create_recommendation_report(user_id=1)
# create_recommendation_report(movie_title='Avatar')
create_recommendation_report()

1. The Shawshank Redemption
   Genres: drama, crime
   Rating: 8.5/10
   Strategy: Trending

2. Fight Club
   Genres: drama
   Rating: 8.3/10
   Strategy: Trending

3. The Godfather
   Genres: drama, crime
   Rating: 8.4/10
   Strategy: Trending

4. Pulp Fiction
   Genres: thriller, crime
   Rating: 8.3/10
   Strategy: Trending

5. The Dark Knight
   Genres: drama, action, crime
   Rating: 8.2/10
   Strategy: Trending

6. Forrest Gump
   Genres: comedy, drama, romance
   Rating: 8.2/10
   Strategy: Trending

7. Inception
   Genres: action, thriller, sciencefiction
   Rating: 8.1/10
   Strategy: Trending

8. Inside Out
   Genres: drama, comedy, animation
   Rating: 8.0/10
   Strategy: Trending

9. Interstellar
   Genres: adventure, drama, sciencefiction
   Rating: 8.1/10
   Strategy: Trending

10. Schindler's List
   Genres: drama, history, war
   Rating: 8.3/10
   Strategy: Trending



,title_x,genres,vote_average,weighted_score,recommendation_type
0,The Shawshank Redemption,"[drama, crime]",8.5,8.548353,Trending
1,Fight Club,[drama],8.3,8.396134,Trending
2,The Godfather,"[drama, crime]",8.4,8.377404,Trending
3,Pulp Fiction,"[thriller, crime]",8.3,8.374738,Trending
4,The Dark Knight,"[drama, action, crime]",8.2,8.344250,Trending
5,Forrest Gump,"[comedy, drama, romance]",8.2,8.272814,Trending
6,Inception,"[action, thriller, sciencefiction]",8.1,8.269290,Trending
7,Inside Out,"[drama, comedy, animation]",8.0,8.256979,Trending
8,Interstellar,"[adventure, drama, sciencefiction]",8.1,8.237399,Trending
9,Schindler's List,"[drama, history, war]",8.3,8.200080,Trending


## Summary:

### Data Analysis Key Findings

*   Informative markdown cells were successfully added before each major section of the notebook: "Load and Prepare Data", "Engineer Features", "Clean Data", "Calculate Weighted Rating", "Building Content-based Recommendation Engine", "Building Collaborative Engine", "Demographic Recommendations", "Hybrid Recommendations", "Evaluate System Performance", "Generate Business Insights", and "Recommendation Report".
*   These markdown cells explain the purpose and rationale behind the code in their respective sections, enhancing the readability and understanding of the notebook.
*   The markdown cells specifically describe the steps involved in data loading and merging, feature engineering (including director, cast, keywords, genres, release year, ROI, and recency), data cleaning and text standardization, weighted rating calculation (incorporating business rules), building content-based and collaborative filtering engines (mentioning SVD and the Surprise library), demographic recommendations (as a cold-start fallback), hybrid recommendation strategy (combining different approaches), system evaluation metrics (Sparsity, RMSE, MAE), generating business insights from the analysis, and the functionality of the recommendation report function.

### Insights or Next Steps

*   The addition of comprehensive markdown explanations significantly improves the documentation and educational value of the notebook.
*   Ensure the markdown cells remain updated if the corresponding code sections are modified in the future.


*   Developed a hybrid recommendation engine leveraging collaborative, content-based, and demographic strategies, achieving a collaborative filtering RMSE of 0.900 on a dataset of 100,000 ratings.
*   Identified "animation", "adventure", and "science fiction" as the highest-revenue generating genres and highlighted 1,421 movies with an ROI > 3x, providing data-driven insights for content acquisition strategy.
*   Analyzed user engagement patterns, finding an average of 149 ratings per user and identifying 22.2% of the user base as highly engaged, informing targeted user retention and engagement initiatives.

###Executive Summary
The entire case study deals with how an airline like United Airlines prepares itself to prevent future aircraft failures by using Machine Learning and operational insights. I used ML techniques on top of a dataset with nearly 50K flight records to develop an advanced predictive maintenance system that accurately predicts the probability of an aircraft failure within 30 days. This enables the airline's maintenance team to quickly identify, analyze and make informed decisions on which flights require pre-emptive inspection or intervention. And this in turn helps in minimizing costly unexpected failures and optimizing maintenance schedules.

* See how it beautifully answers all the questions in one flow:
  * ✅What problem did I try to solve?
  * ✅What type of data I used?
  * ✅How I did it?
  * ✅What will potentially happen to the respective team after I did it?

###Business Problem (Why it matters)
For any airline, preventing unexpected aircraft failures is extremely important for safety, operational efficiency, and profitability. If the airline can predict potential failures before they happen, they can schedule maintenance proactively, avoiding costly delays, cancellations, and most importantly, ensuring passenger safety. This problem impacts passengers, maintenance and operations teams, and the airline's overall financial performance and reputation. Building a good predictive maintenance system helps to reduce downtime, lower maintenance costs, improve aircraft utilization, and enhance safety records. If this problem isn't handled well enough, unexpected failures would continue, leading to significant operational disruptions, increased repair costs, and potential safety risks, damaging the airline's credibility and bottom line.

* See how it beautifully answers all the questions in one flow:
  * ✅Why solve this problem and how it matters to the business?
  * ✅Who's affected by this problem?
  * ✅How this solution saves money, reduces risk, improves efficiently?
  * ✅What would happen if this wasn't solved?

###My approach (Story of the build)
Let's walkthrough the approach:

Took the airline maintenance data → cleaned and explored features, noting the imbalance in failure data → looked at flight info, sensor data, environmental factors, and maintenance history → created new features like 'temp_delta', 'stress_score', and 'maintenance_risk' to capture degradation trends and interaction effects → sorted data by aircraft and flight hours to calculate trend features accurately → scaled numerical features and used Target Encoding for categorical ones → split data into train-test sets, ensuring stratification for the imbalanced target → computed class weights to handle imbalance during model training → tried out several models (Logistic Regression, XGBoost, LightGBM, Isolation Forest) → evaluated them based on ROC-AUC and Precision-Recall AUC scores (better for imbalanced data) → Logistic Regression with engineered features had the best PR-AUC and ROC-AUC, outperforming tree-based models and anomaly detection, indicating failures follow mostly linear degradation patterns → Statistical performance metrics that performed to me were: ROC-AUC and PR-AUC (for imbalanced classification), Confusion Matrix (to see False Positives and False Negatives), Classification Report (Precision, Recall, F1-score) → Analyzed feature importance for insights (Logistic Regression coefficients, XGBoost/LightGBM importance) → Calculated business impact based on defined costs for false positives and false negatives → Included cost-sensitive evaluation to find optimal threshold balancing business costs → Prepared executive-level summary and roadmap.

* See how it beautifully answers all the questions in one flow:
  * ✅How did I clean and understand the data?
  * ✅What features did I use/create and why?
  * ✅What models did I try and why did I pick the final one?
  * ✅What metrics mattered to me and the business?
  * ✅How did I handle edge cases like imbalance or drift?

###Model Interpretation (What did I learn?)
Yes, I can explain the working of this predictive maintenance model to a non-technical stakeholder. Basically, we've built a system that helps predict whether an aircraft is at risk of failure within the next 30 days. Think of it like a smart assistant that looks at different pieces of information about a flight and the aircraft's history and gives either a 'green light' (low risk) or a 'red flag' (high risk). How it does that? Essentially, it's using historical data of past flights and maintenance records to learn exactly what characteristics and trends are associated with aircraft failures and which ones are not.
Some factors that stood out were:
Flight hours and cycles (cumulative wear and tear); Sensor data like engine temperature and vibration level (indicators of system health); Environmental/Route factors like turbulence and airport condition (external stress); Maintenance history (recency and quality of last check); Trends and deltas in sensor readings (capturing degradation over time); Combined stress scores (interaction of vibration, turbulence, and airport condition); Maintenance risk score (interaction of days since last check and its quality). Yes, the model's predictions appear logical from a business POV. Why? Analysis shows that the model can significantly reduce expected maintenance costs by flagging high-risk flights for pre-emptive checks. The model focuses on factors known to contribute to mechanical stress and wear. Since the model performed well on test data and its feature importance aligns with domain knowledge, it suggests it learned general patterns of failure rather than just memorizing training examples. For handling bias, we used stratification during the train-test split and class weights/imbalance handling in models to ensure the model doesn't ignore the rare failure cases during training.

* See how it beautifully answers all the questions in one flow:
  * ✅Can I explain the working of this model to a non-technical stakeholder and how do they work?
  * ✅Which features were most important and why?
  * ✅Are the model's predictions logical from a business POV?
  * ✅How did I validate that its not overfitting or biased?

###Business Value
Quantifying the business impact of this model:
* Expected cost with model (threshold 0.4-0.5): ~$4.29M - $4.30M per 50k flights
* Baseline 'do nothing' cost: ~$28.8M per 50k flights (5.76% failure rate * $10k cost per failure)
* Cost reduction: ~85% reduction in expected losses compared to baseline.
* Recall at threshold 0.4: ~86% (catching most failures)
* Precision at threshold 0.4: ~6.6% (manageable false alarms for safety-critical use)

Some of the KPIs relevant to the business purpose were:
Expected Cost (directly measures overall financial impact considering false positives and negatives); Recall (critical for safety - percentage of actual failures correctly identified); Precision (important for operational efficiency - percentage of flagged flights that actually fail); False Positive Rate (cost of unnecessary inspections); False Negative Rate (cost and safety risk of missed failures); ROC-AUC and PR-AUC (overall model discrimination power, especially for rare events).

* See how it beautifully answers all the questions in one flow:
  * ✅Can you quantify the business impact of your model?
  * ✅What KPIs are relevant to the business purpose?

###Future Enhancements (What I'd improve)
If I had more time/data/resources, then I'd definitely work on certain aspects:

Advanced Data/Features → Integrate real-time sensor data streams; Engineer time-series features (e.g., rate of change, cumulative stress cycles); Explore external data sources (e.g., detailed flight logs, specific part histories).

MLOps Integration → Implement automated data pipelines for continuous model retraining; Set up A/B testing for different model versions or thresholds; Develop model monitoring for drift detection (input data characteristics, prediction distributions).

Product Expansion → Apply models to specific critical components (engines, landing gear); Develop prescriptive maintenance recommendations based on failure type; Integrate into fleet management systems for dynamic scheduling.

* See how it beautifully answers all the questions in one flow:
  * ✅What would I do if I had more time/data/resources?

###Monitoring & Maintenance
If the model's performance degrades after it's put into use, it means it's not as good at predicting aircraft failures as it was when we tested it. This results in increased unexpected failures, higher maintenance costs, reduced operational reliability, and potential safety risks, harming the airline's reputation and profitability. Need to keep a track of the key business metrics (KPIs discussed earlier) like unexpected failure rate and maintenance costs. Monitor for data drift, continuously checking if the characteristics of incoming flight and sensor data change significantly. Monitor for prediction drift, tracking how the model's risk scores and predictions are distributed over time. Analyze model explanations periodically to ensure the model is still making decisions based on logical risk factors and that feature importance remains consistent. A/B testing, if possible, comparing the current model with updated versions or alternative approaches on a subset of flights.

* See how it beautifully answers all the questions in one flow:
  * ✅What happens if your model performance degrades?
  * ✅How would you monitor your model post deployment?

###Deployment
This project wasn't deployed. In order to do that, the model and code need to be finalized and packaged (e.g., via Docker container) → set up automated data pipelines (ingesting real-time or near-real-time flight and sensor data to produce risk scores) → build APIs (creating endpoints for the model to be accessed by other systems like maintenance scheduling software or operational dashboards) → host the APIs on cloud platforms (like AWS, Azure, or GCP) → integrate the deployed API with United Airlines' existing operational systems (fleet management, maintenance planning, pilot reporting tools).

###Reflections & Takeaways
Good models make predictions, great models provide actionable insights – being able to explain *why* a flight is high-risk helps maintenance teams target inspections effectively.

It's not just about building a model with high statistical scores – it's about building a solution that solves a specific business problem and provides tangible value, so understanding the operational context and cost implications is crucial.

Simple models can be powerful – sometimes a well-engineered linear model captures the core dynamics better than complex non-linear ones, offering interpretability crucial for business adoption.

* See how it beautifully answers all the questions in one flow:
  * ✅What did you takeaway from this entire project?

###Executive Summary
This case study focuses on developing a comprehensive recommendation system for Netflix, designed to enhance user engagement and satisfaction by providing personalized movie suggestions. Leveraging a dataset containing movie metadata and user ratings, the project employs a hybrid approach combining collaborative filtering, content-based filtering, and demographic recommendations. The system accurately predicts user ratings and suggests movies based on individual preferences and movie characteristics, thereby minimizing user effort in discovering new content and optimizing content consumption.

* See how it beautifully answers all the questions in one flow:
  * ✅What problem did I try to solve?
  * ✅What type of data I used?
  * ✅How I did it?
  * ✅What will potentially happen to the respective team after I did it?

###Business Problem (Why it matters)
For a streaming service like Netflix, providing relevant and engaging content recommendations is crucial for retaining subscribers and increasing viewing time. Ineffective recommendations can lead to user frustration, difficulty in finding enjoyable content, and ultimately, churn. This problem impacts subscribers (poor viewing experience), content strategists (missed opportunities for content promotion), and the company's overall financial performance and market position. A well-designed recommendation system helps to personalize the user experience, drive content discovery, increase session duration, and improve subscriber satisfaction and retention. If this problem isn't handled well enough, users would struggle to find content they like, leading to decreased engagement, higher churn rates, and a negative impact on Netflix's growth and profitability.

* See how it beautifully answers all the questions in one flow:
  * ✅Why solve this problem and how it matters to the business?
  * ✅Who's affected by this problem?
  * ✅How this solution saves money, reduces risk, improves efficiently?
  * ✅What would happen if this wasn't solved?

###My approach (Story of the build)
Let's walkthrough the approach:

Loaded and prepared the movie metadata and user ratings datasets → merged the datasets based on common identifiers → engineered features by parsing JSON strings for cast, crew, keywords, and genres, extracting relevant information like director and top cast members → created additional features such as release year, Return on Investment (ROI), and a recency indicator → cleaned and standardized text features (cast, keywords, director, genres) to ensure consistency → built a "metadata soup" by combining relevant text features to capture movie content characteristics → calculated weighted ratings for movies, incorporating business rules like minimum vote count and boosting scores based on popularity and recency → developed a content-based recommendation engine using `CountVectorizer` and cosine similarity on the metadata soup → built a collaborative filtering engine using the Surprise library and the SVD algorithm on user ratings, including a train-test split for evaluation → implemented a demographic recommendation fallback based on weighted scores for cold-start users or when other methods are insufficient → created a hybrid recommendation function that combines collaborative, content-based, and demographic approaches based on user history and movie interest → evaluated the system's performance using metrics like sparsity, RMSE, and MAE for collaborative filtering → generated business insights by analyzing popular genres, high-ROI movies, and user engagement patterns → finally, created a recommendation report function to present recommendations in a user-friendly format.

* See how it beautifully answers all the questions in one flow:
  * ✅How did I clean and understand the data?
  * ✅What features did I use/create and why?
  * ✅What models did I try and why did I pick the final one?
  * ✅What metrics mattered to me and the business?
  * ✅How did I handle edge cases like imbalance or drift?

###Model Interpretation (What did I learn?)
Yes, I can explain the working of this movie recommendation system to a non-technical stakeholder. Basically, we've built a system that suggests movies you might like based on different factors. Think of it like having a personalized movie expert. How it does that? It uses three main ways:
1. **Looking at what movies are popular and well-rated:** It identifies movies that many people like and have high scores, recommending these as a general fallback.
2. **Looking at what movies are similar to ones you like or are interested in:** It analyzes the content of movies (like actors, directors, genres, keywords) to find others that share similar characteristics.
3. **Looking at what other people with similar tastes liked:** It uses patterns from how many users rated movies to predict what you might rate movies you haven't seen yet.

Some factors that stood out were: Movie genres and keywords (help understand the content), cast and director (influence movie style and appeal), user rating history (reflects individual preferences), and movie popularity/vote counts (indicate broad appeal). Yes, the model's predictions appear logical from a business POV. Why? The system aims to increase user engagement by showing relevant content, which aligns with Netflix's goal. By combining different approaches, it can handle various scenarios, like recommending to new users or suggesting movies based on a specific interest. The evaluation metrics (like RMSE for collaborative filtering) show that the model can reasonably predict user ratings, suggesting it has learned underlying patterns in the data. For handling challenges like cold start (new users or movies), the demographic and content-based methods provide recommendations even without extensive rating history.

* See how it beautifully answers all the questions in one flow:
  * ✅Can I explain the working of this model to a non-technical stakeholder and how do they work?
  * ✅Which features were most important and why?
  * ✅Are the model's predictions logical from a business POV?
  * ✅How did I validate that its not overfitting or biased?

###Business Value
Quantifying the business impact of this model:
* **Increased User Engagement:** By providing relevant recommendations, users spend less time searching and more time watching, leading to higher session duration.
* **Improved Subscriber Retention:** Personalized experiences contribute to higher user satisfaction, reducing churn rates.
* **Optimized Content Discovery:** The system helps users find hidden gems or content they wouldn't have otherwise discovered, increasing the utilization of the content library.
* **Informed Content Strategy:** Insights from genre analysis and ROI can inform decisions about acquiring or promoting certain types of content.

Some of the KPIs relevant to the business purpose were:
Recommendation Click-Through Rate (CTR) (measures how often users click on recommended movies); Average Session Duration (indicates how long users watch content per session); Subscriber Churn Rate (measures the percentage of subscribers who leave the service); Content Discovery Rate (tracks the proportion of content viewed that was recommended); User Satisfaction Scores (gauged through surveys or feedback mechanisms); Diversity of Recommendations (ensuring users are exposed to a variety of content).

* See how it beautifully answers all the questions in one flow:
  * ✅Can you quantify the business impact of your model?
  * ✅What KPIs are relevant to the business purpose?

###Future Enhancements (What I'd improve)
If I had more time/data/resources, then I'd definitely work on certain aspects:

Advanced Modeling Techniques → Explore deep learning models (e.g., neural collaborative filtering, sequence-aware models) for potentially better prediction accuracy; Implement matrix factorization techniques beyond SVD (e.g., FunkSVD, BiasSVD).

Real-time Recommendations → Develop a system for generating recommendations in real-time based on user interactions within a single session; Incorporate contextual information (time of day, device, mood) into recommendations.

Explainability and Trust → Implement techniques to provide explanations for why a movie was recommended, increasing user trust and understanding; Allow users to provide feedback on recommendations to further personalize the system.

A/B Testing and Optimization → Set up a robust A/B testing framework to compare different recommendation algorithms and strategies; Continuously optimize recommendation parameters based on key business metrics (e.g., CTR, watch time).

* See how it beautifully answers all the questions in one flow:
  * ✅What would I do if I had more time/data/resources?

###Monitoring & Maintenance
If the model's performance degrades after it's put into use, it means it's not as good at recommending movies as it was when we tested it. This results in users seeing less relevant suggestions, potentially leading to decreased engagement and higher churn. Need to keep a track of the key business metrics (KPIs discussed earlier) like recommendation CTR and average session duration. Monitor for data drift, continuously checking if the characteristics of incoming user behavior or new movie data change significantly. Monitor for prediction drift, tracking how the model's recommended ratings or similarity scores are distributed over time. Analyze recommendation diversity periodically to ensure the system isn't recommending the same popular movies to everyone. A/B testing, if possible, comparing the current model with updated versions or alternative approaches on a subset of users.

* See how it beautifully answers all the questions in one flow:
  * ✅What happens if your model performance degrades?
  * ✅How would you monitor your model post deployment?

###Deployment
This project wasn't deployed. In order to do that, the model and code need to be finalized and packaged (e.g., via microservices) → set up automated data pipelines (ingesting real-time user interaction and content data) → build APIs (creating endpoints for the recommendation engine to be accessed by the Netflix application) → host the APIs on cloud platforms (like AWS, Azure, or GCP) with considerations for scalability and low latency → integrate the deployed API with Netflix's user interface and backend systems to display recommendations to users.

###Reflections & Takeaways
Understanding the nuances of different recommendation approaches (collaborative, content-based, demographic) and when to apply them is key to building a robust hybrid system.

Data cleaning and feature engineering, especially parsing complex nested data structures, are critical steps that significantly impact the performance of content-based methods.

Balancing statistical performance metrics (like RMSE) with business-relevant KPIs (like CTR and retention) is essential for creating a system that is not only accurate but also drives tangible business value.

Addressing cold-start problems is crucial for the success of a recommendation system, and incorporating fallback strategies like demographic recommendations is a practical solution.

* See how it beautifully answers all the questions in one flow:
  * ✅What did you takeaway from this entire project?